### ISSUE TITLE
Allow optional scoping for including a launchfile

### ISSUE THREAD

**Issue by SuperJappie08** [ros2/launch#801]
## Feature request
`IncludeLaunchDescription` is not scoped if I understand correctly. (Based on #313 #689 and [the design docs](https://design.ros2.org/articles/roslaunch.html)

#### Feature description

An extra argument for `IncludeLaunchDescription` action, which could enable scoping.
This would prevent the leakage of arguments.

#### Implementation considerations

This is already possible by including into a group with `forwarding` set to false.
Default could be false to not break the current behavior.

**Comment by wjwwood** [comment]
@SuperJappie08 thanks for the contribution, it's definitely something that others have run into and found surprising.

I really like the idea of a new Action, like `ScopedIncludeLaunchDescription` or similar, and I would be in favor of including it in this repository if you want.

Also, I really appreciate you making it into a stand-alone package as a first step. It definitely helps demonstrate the idea and value, and it makes it easier for folks to try it or use it, even in older releases of ROS 2.

**Comment by alsora** [comment]
Hi @SuperJappie08 any update on this?

**Comment by SuperJappie08** [comment]
> Hi [@SuperJappie08](https://github.com/SuperJappie08) any update on this?

Hi,

I was initially waiting on a response from a maintainer before spending a lot of time on this because I have a tendency to spend a lot of time on the wrong things.
However, some family stuff has come up recently, which required a shuffle in my priorities, so I have not had the time (yet) to start this.

I think I would be able to start this week.

**Comment by Timple** [comment]
>  it's definitely something that others have run into and found surprising.

Certainly!

So why not make this the default? Breaking changes are allowed between ROS releases. 

I know breaking backward compatibility should not be considered lightly. But it's a trade-off between:
- new developers running into this surprising feature
- existing developers actually using this feature

The first group might actually be larger than the second.

**Comment by SuperJappie08** [comment]
> > it's definitely something that others have run into and found surprising.
> 
> Certainly!
> 
> So why not make this the default? Breaking changes are allowed between ROS releases.
> 
> I know breaking backward compatibility should not be considered lightly. But it's a trade-off between:
> 
>     * new developers running into this surprising feature
> 
>     * existing developers actually using this feature
> 
> 
> The first group might actually be larger than the second.

Having it as a separate action, could mean the default include action could change.
But I think your point is valid, and it possibly could make sense to rename the old `IncludeLaunchDescription` to `IncludeUnscopedLaunchDescription` so the new scoped version could be the easier/shorted(/ more likely to be adopted as the default) name (`IncludeLaunchDescription`).

Normally, I would also cheer for backwards compatibility, but since the unscoped version feels counterintuitive to many people, it might make sense to _fix_ it.

EDIT: Clarification

**Comment by SuperJappie08** [comment]
I found a fatal flaw after doing more testing on this feature with an implementation similar to my initial attempt in the package.

Currently, the 'scoped' is essentially a normal include wrapped in a scoped non-forwarding group, which only keeps the provided launch arguments.
This gives essentially the expected behavior from this feature, however, since **only** the launch argument to the nested files are kept in systems depending on non-user set launch configurations, such as the `PushRosNameSpace` and `SetRemap` actions from `launch_ros`.

My initial reaction was to attempt to include launch configurations which were not defined explicitly by the child file. 
However, this requires a lot of rework to support multi-nesting situations.

Since not being able to do things, like name spacing a launch file, would not resolve the current non-scoping confusion behavior, I am unsure on how to continue (for now).

Does anyone have suggestions for a direction to move forward? (I will try to think of reasonable solutions as well)

EDIT: My current suggestion would be to pass all launch configurations except those explicitly defined in (recursively) included files. However, this does not seem a clear and understandable solution to users of the scoped include action

**Comment by SuperJappie08** [comment]
Another option could be to register overscoping launch configurations in the context, which will always be included.
This could be implemented as a launch configuration, which contains the keys of the unscopable launch configurations.

**Comment by SteveMacenski** [comment]
I'd love to have something like this in the package (and by default!) :-)

**Comment by SuperJappie08** [comment]
@alsora, an initial working implementation is complete (#841).

**Comment by ijnek** [comment]
> This gives essentially the expected behavior from this feature, however, since only the launch argument to the nested files are kept in systems depending on non-user set launch configurations, such as the PushRosNameSpace and SetRemap actions from launch_ros.

So, I think the concerns are these launch configurations:
* ros_remaps
* ros_namespace
* global_params

I think a nice solution would be to have no launch configurations forwarded as done in this PR, and have a IncludeLaunchDescription (or something named a bit different to avoid confusion - like RosIncludeLaunchDescription), that inherits the ScopedIncludeLaunchDescription here, but auto-forwards the launch configurations listed above.

This can be thought of a similar concept to the launch_ros' ``Node`` class extending launch's ``ExecuteProcess``.

**Comment by SuperJappie08** [comment]
@ijnek Thank you for the input.
That is an option I hadn't considered yet.

However, my concern is that a ROS specific `RosIncludeLaunchDescription` is less flexible, since in the current attempted implementation, other launch extensions (think a [`launch_system_nodes`](https://index.ros.org/p/launch_system_modes/) or something like a (non-existing) `launch_gz`) could just hook/register their own exclusions. 
This would make combing extensions trivial, whereas with the alternative, this would get more complex.

However, it is a balancing act, since these exclusions should not be overused and are _only practical_/_make sense_ in specific situations.

I'm unsure what the best way forward would be in this case, where it's a choice between flexibility vs. simplicity.

**Comment by ijnek** [comment]
Sorry — I completely missed your comment (https://github.com/ros2/launch/pull/841#issuecomment-2845561097) mentioning that you had already worked around this issue using global launch configurations.

Personally, I find it odd that we use launch configurations to pass ros_remaps, ros_namespace, and global_params to included launch files. While looking into launch, it seems we could instead store remapping and namespacing info in the local stack rather than using launch configurations.

https://github.com/ros2/launch/blob/820e5a2c57dafd807ea4cdf33952e87c78929ce5/launch/launch/launch_context.py#L63-L64

I’ve drafted a PR to explore this idea: https://github.com/ros2/launch_ros/pull/470.

This approach would avoid the need to track a global list of which launch configurations are meant to propagate. Let me know what you think.

**Comment by SuperJappie08** [comment]
That would indeed also resolve those weird behaviors.

Personally, I'm not set to on the global exclusion list, since it is unclear in the internal API, since there are also global variables in the launch context. So naming and behavior of the global list is not very transparent in my opinion.

Currently, I'm mostly curious, why these `launch_ros` features use the launch configurations instead of launch context locals and why that decision was made in the first place.
Because if it was a conscious decision to do it like this, there might be a valid reason we are currently missing.

EDIT: After some more looking into the launch internals, scoping local variables is rarely used, currently it is only used in the `ForEach` action (also tests and event internals). However, this would be easy to integrate with the scoped include.
However, that would mean that ros2/launch_ros#470 does not work with the normal group action anymore.

**Comment by emersonknapp** [comment]
Discussed this in ROSGraph WG today, we think this issue is important and covers a lot of confusing behavior in launch, but needs a design doc / summary pass before diving back into implementations, so that we can discuss at that level

**Comment by SuperJappie08** [comment]
> Discussed this in ROSGraph WG today, we think this issue is important and covers a lot of confusing behavior in launch, but needs a design doc / summary pass before diving back into implementations, so that we can discuss at that level

I will do this somewhere in the coming weeks.

**Comment by SuperJappie08** [comment]
Weeks turned into months, but I have time now.

The full summary:

# Goal
The purpose of scoped include is to simplify the inclusion of (child) launchfiles without exposing the inner workings of them, while not effecting the parent launchfile.
This would prevent the non-passed launch arguments/configurations from being exposed to the inner child launch file.
In short, the child cannot affect the parent and only has access to the explicitly exposed launch parameters.

This was already possible with a combination of the `group` and `include` actions, however it is cumbersome, as all the passed-arguments/configurations need to be passed defined twice, ones in the `Group` action (as a `keep` attribute) and onces to the `IncludeLaunchDescriptionAction` via the `arg`/`let` attribute.

Inititally, I implemented the `ScopedIncludeLaunchDescription` as combination of these to Actions, which does this extra passing internally. 

# The Problem
However, during the testing of this, I found that applying this strict scoping to a launchfile results in unclear behavior for the users.
As actions (from `launch_ros`) such as `PushROSNamespace`, `SetParameter` (global) and `SetRemap`  don't work with the included launchfile, as they store their information in the `launch_configurations` attribute of the `LaunchContext`.

Not being able to perform these operations on the included launch file makes this feature less practical and confusing to beginners.
(as the 'workaround' would be to pass these internal launch configurations, such as `ros_namespace` to the scoped include action)

# Initial Mitigation
As an initial mitigation (which I'm not happy with), I made an scoping exclusion list.
This is really flexible, as extensions can '_register_' their unscoped arguments at runtime, which allows it to be adopted widely.

However, I think this system of global launch configurations is non-transparent, difficult to grasp, and introduces unnecessary complexity.
Furthermore, it can be easily 'abused', making this feature behave unpredictably.

# The new proposed solution
A better, but more invasive solution, would be to migrate these internal launch configurations over to the `locals` of the launch context.
This would require that the `GroupAction` also scopes the locals (which could be affected by the existing flags, `'scoped'` and `'forwarding'`).
The scoped include action would then still forward the locals into its scope, which ends after the launchfile, resulting in a much clearer behavior.

The main downside of this would be that it would probably break some existing launchfiles/extensions.
It would also mean that the ROS namespace now needs to be accessed via a purpose-built substitution instead of via the launch argument.

Personally, I think the new solution is the most promising; however, I want **explicit green-light** as this will definitely break something somewhere.

EDIT: The greenlight would be to confirm I'm atleast on the right track, as I will then also try and find the effected launch files and extensions.

**Comment by paulsohn** [comment]
Hi @SuperJappie08 ,
I suffer from the same issue, so I am developing my own launcher linter, and I can't wait to see that all includes are forced to explicitly specify either `scoped=True` or `scoped=False` (at least being warned by not setting it). My question is: why doesn't this simple proposal work?

I feel like PushROSNamespace or SetParameter are already propagated even with the existing `GroupAction(scoped=True)`, and I think it's worthy to inspect on how GroupAction actually handles scoping, which is what I'm currently working on.

**Comment by paulsohn** [comment]
I have reviewed the current status and implemented my patch. 

> This is already possible by including into a group with `forwarding` set to false.

Was discussion on `forwarding=False` necessary? I believe our desired solution is to make `<include scoped="true" (...) />` equivalent to `<group scoped="true"><include (...) /></group>`, so that the includee can see all the ROS namespaces and params inherited from the includer, allowing it cloberring all the environment and context without propagating to the includer.

I think this discussion is irrelevant to `forwarding`, in which case we could've implemented the fix earlier. Still, even on top of #953 we can add `forwarding` support analogously how we do for `GroupAction` - but I will leave it as a future work.

**Comment by Timple** [comment]
We could put up a discourse post asking for feedback. We can try this already on some large public code bases like autoware.

If there was some kind of dummy implementation of some easy utility to test this, people could easily evaluate their own internal projects.
If nobody is really affected this would make a strong argument for making it a default.

**Comment by paulsohn** [comment]
@Timple I am working for a company using Autoware, and at this moment Autoware is more than a test bed and it's actually used by some real-world vehicle systems. Plus autoware_launch is *deliberately* using scope leakage in their recent commits which haven't passed a day from this reply. (This is also mentioned in the description of #953 .)

So I am against making a breaking change at once - it will only introduce more friction against introducing the necessary and long-waited feature to support scoped include in any form. Instead we can add warning for includes not explicitly specifying `scoped=True/False`.

**Comment by Timple** [comment]
> If nobody is really affected this would make a strong argument for making it a default.

> autoware_launch is deliberately using scope leakage

My statement of course also works the other way around. This makes it a strong argument for not making it a default!

**Comment by SuperJappie08** [comment]
I think the main difference with this original attempt compared to #953 is that my attempt also blocks not explicitly passed launch configurations, causing the namespacing issue.
However, this extra blocking prevents the included launch files from being affected by the parent (which can be useful, for example, for external modules.).

> Plus autoware_launch is _deliberately_ using scope leakage in their recent commits which haven't passed a day from this reply. (This is also mentioned in the description of [#953](https://github.com/ros2/launch/pull/953) .)

To my understanding, this behavior of _leaking_ is actually intended behavior, as normal `include` is a flat copy-past style action.
(Which can be really useful from time to time.)

**Comment by paulsohn** [comment]
> To my understanding, this behavior of *leaking* is actually intended behavior, as normal `include` is a flat copy-past style action.

Agreed, assuming that if the user is always sure what they're doing. Autoware is suffering from unintended leaks.

I think it is sufficient and the clearest for having:
* `<include scoped="true/false" (...) />` should directly correspond to `<group scoped="true/false"><include (...) /></group>`
* `<include scoped="true" forwarding="true/false" (...) />` should directly correspond to `<group scoped="true" forwarding="true/false"><include (...) /></group>`

and I feel we better not have a design that `<include scoped="true" (...) />` sets up both scoping and blocking forwarding. 
If we want it, we can add `forwarding` attribute separately on top of #953 .

**Comment by SuperJappie08** [comment]
Sorry for responding in weird order.

> Hi [@SuperJappie08](https://github.com/SuperJappie08) , I suffer from the same issue, so I am developing my own launcher linter, and I can't wait to see that all includes are forced to explicitly specify either `scoped=True` or `scoped=False` (at least being warned by not setting it). My question is: why doesn't this simple proposal work?

In my implementation, it is not a toggle, but a separate action to avoid confusion. 
It should always be an optional opt-in style behavior.

The reason this doesn't for this specific case is that it will still pass non-specified arguments, causing defined optional arguments (with a default) in the innerscope to be overwritten unintentionally if the launch configuration name is also used in the parent before the include.

EDIT: This written before the previous message was received.

**Comment by SuperJappie08** [comment]
> I think it is sufficient and the clearest for having:
> 
>     * `<include scoped="true/false" (...) />` should directly correspond to `<group scoped="true/false"><include (...) /></group>`
> 
>     * `<include scoped="true" forwarding="true/false" (...) />` should directly correspond to `<group scoped="true" forwarding="true/false"><include (...) /></group>`
> 
> 
> and I feel we better not have a design that `<include scoped="true" (...) />` sets up both scoping and blocking forwarding. If we want it, we can add `forwarding` attribute separately on top of [#953](https://github.com/ros2/launch/pull/953) .

I think this might be oversimplifying the issue, as the scoping/forwarding behavior has a relatively unclear meaning.

Also, see the comment above about unintentional argument overwriting.

I think a separate action with constant behavior would greatly benefit the system since it prevents complex edge cases.
These complex cases where other behavior is required can already be solved by nesting a group and include as mentioned.

----
As a clarification side note, with scoped include, I mean more than just the `group` actions scope attribute, as I'm more thinking of an isolated launch scope.

**Comment by paulsohn** [comment]
> Also, see the comment above about unintentional argument overwriting.

I argue that for your requirements we can still use `<include scoped="true" forwarding="false" (...) />` from my suggestion. Currently I only added `scoped` in #953 but adding `forwarding` is just analogous. Do you want me to implement in the same PR?

**Comment by SuperJappie08** [comment]
> > Also, see the comment above about unintentional argument overwriting.
> 
> I argue that for your requirements we can still use `<include scoped="true" forwarding="false" (...) />` from my suggestion. Currently I only added `scoped` in [#953](https://github.com/ros2/launch/pull/953) bud adding `forwarding` is just analogous. Do you want me to implement in the same PR?

The thing that I failed to mention is that that was what I was basically doing. (so you can add it if you like).
However, it does not solve the issue fully.

(To avoid confusion, I will switch terminology to isolated include for my intended behavior)

As pushing a namespace and isolated including won't work as that internally uses the launch configurations, which then would have to be passed explicitly.
This is confusion, as I doubt it is widely known which launch configurations you have to forward to get the correct behavior.

**Comment by gavanderhoorn** [comment]
@SuperJappie08: would it perhaps be possible to create some tests that show the (un)desired behaviour?

Would also perhaps be less ambiguous than natural language descriptions.

**Comment by paulsohn** [comment]
If my suggestion does not guarantee the 'complete isolation', then it is highly likely that neither does the current `GroupAction(scoped=True, forwarding=False)`. How about suggesting what we can do for completely isolated `GroupAction` as well?

**Comment by SuperJappie08** [comment]
> [@SuperJappie08](https://github.com/SuperJappie08): would it perhaps be possible to create some tests that show the (un)desired behaviour?
> 
> Would also perhaps be less ambiguous than natural language descriptions.

@paulsohn. I'm writing a package to demonstrate the current behavior, which includes some introspection magic.
This will take a bit longer, will reply again once I have it working.

**Comment by paulsohn** [comment]
Speaking from my gut feeling, shouldn't the 'isolated including' imply *not* propagating the caller namespaces as well?
Anyways, maybe more granular filtering such as `forwarding="ros_namespace"` instead of boolean, or explicit context space separation (which is unrealistic because it requires the entire rewriting of launch / launch_ros) might help for your desired usecase.

I now see that the goal of this issue is beyond what I implemented in #953 , but nevertheless that PR has its own importance and I believe it is also a prerequisite before closing this issue.

**Comment by SuperJappie08** [comment]
@paulsohn, I've made the demo package.

https://github.com/SuperJappie08/isolated-launch-demo

INSERTED-EDIT: I've opted to emulate node namespace behavior instead of starting actual nodes to keep the output clearer.

> Speaking from my gut feeling, shouldn't the 'isolated including' imply _not_ propagating the caller namespaces as well? Anyways, maybe more granular filtering such as `forwarding="ros_namespace"` instead of boolean, or explicit context space separation (which is unrealistic because it requires the entire rewriting of launch / launch_ros) might help for your desired usecase.

The issues I could see occuring with a forwarding filter are:
 - This cannot be a preset filter, as there could be any number extentions that introduces more of these internal configurations.
 - If this was a list of configurations to forward, the user needs to know these configuration names, which are internal so they are not well known. 

---

@paulsohn, The last demo shows a setup that will also break #953, if `forwarding` is introduced.

**Comment by SuperJappie08** [comment]
To reiterate from earlier, I would suggest migrating such internal launch configurations to launch locals for most extensions, so those can be scoped independently from the _'user'_ launch configurations.

**Comment by paulsohn** [comment]
1. #953 currently is irrelevant to `forwarding`.
2. Isn't it breaking just because you wanted to define `<keep>` in your own way? I can argue that it is YOUR `<keep>` that breaks and my `forwarding="filter"` will work regardless, though I currently have no plan to work further on supporting forwarding (either selective or total).

> If this was a list of configurations to forward, the user needs to know these configuration names, which are internal so they are not well known.

But this also applies to your `<keep name=(...)>`. PushROSNamespace is actually a special launch entry that it has its own hardcoded key, so I don't see that using `ros_namespace` is touching 'internal details'.

**Comment by SuperJappie08** [comment]
> But this also applies to your `<keep name=(...)>`. PushROSNamespace is actually a special launch entry that it has its own hardcoded key, so I don't see that using `ros_namespace` is touching 'internal details'.

The point I'm trying to make is that there is no convention for this, as `ros_namespace` is not the only example for such an _internal_/not-explicitly-user-set configuration.

The _fix_ I suggest in the repository is meant as current work-around for the issue, as that would be the way to do it without any launch updates.

There are ways to solve the isolation issues, however these 'special' cases (e.g. namespace, global remaps etc.) are difficult to deal with in the current situation. 

Being able to namespace these isolated includes, allows for behavior similar to ROS 1, as well as simplifying a lot of 'bringup'-style launch files (in which multiple subsystem launchfiles get started)
This use-case was also (shortly) mentioned during a ROSGraph working group meeting, under which banner I attempted to finish this.

A solution for this is to change how namespace pushing, etc. work by utilizing the local scope of the launch context to store this instead.
Locals already get scoped separately, as they already persist over all possible group action configurations.
This could allow isolation to work fully with a minor adaptation off the original start of #841 ([The first and second commit](https://github.com/ros2/launch/pull/841/changes/6220d1bc83cce1675c51176b0d502024efdb81b1))

**Comment by paulsohn** [comment]
I would say that this issue thread tries to handle two problems without clear separation -- one is outward (launch configuration from the included launch file pollutes the outer launch file) and one is inward (launch configuration set from outside is exposed to the included launch file).

**Comment by SuperJappie08** [comment]
> I would say that this issue thread tries to handle two problems without clear separation -- one is outward (launch configuration from the included launch file pollutes the outer launch file) and one is inward (launch configuration set from outside is exposed to the included launch file).

I think you are correct.
Initially, I saw it as one problem, but it now seems to make more sense as 2 separate cases.

**Comment by wjwwood** [comment]
Sorry for the late reply, I've been searching for time to catch up on the conversation and weigh in. 

I'm gonna tl;dr my opinions here and then follow with details for my rationale:

- `IncludeLaunchDescription` being unscoped was intentional
  - I find it helpful to consider launch files as markup and not code, and so it was intentionally made to resemble systems that modify text, not the semantics of code
  - it resembles C++'s preprocessor `#include` statement, i.e. an include statement includes the content in-place as if it were literally written there, but not the C++ compiler which considers the semantics of the content
  - also resembles most templating engine's include statements (most are not scoped in my experience)
  - as opposed to CMake's include system where you must explicitly use PARENT_SCOPE to write to the parent scope
  - this behavior makes sense (in my opinion) when breaking up a launch file for a single task into multiple files, or when reusing fragments in multiple places
  - this is surprising behavior (in my opinion) when making top-level launch files that include unrelated launch files (each specific to an entire task) together
    - the idea was that top-level launch files would use scoped group actions to resolve this
- `IncludeLaunchDescription` including the parent scope and also taking `launch_arguments` was intentional
  - again this behavior (including the parent scope by default) makes sense if you consider it to be a parallel to C++'s `#include` statement or a template engine's include
  - the purpose of `launch_arguments` was to explicitly check that required launch arguments in the included launch file were provided, and to produce earlier feedback when you failed to provide them
  - you can do basically everything without `launch_arguments` since the parent scope is there, but speculative error checking was difficult without it, however I think there's an argument to made that this argument was a mistake, as it mixes content manipulation (including the sub-launch description) and semantics (were the require arguments of the sub-launch description satisfied)
  - `launch_arguments` also help you to be explicit when remapping arguments from the parent scope, but that could also be done with `SetLaunchConfiguration`, e.g. when the parent launch file is multiplexing options to sub launch files like `<include ...><let name="name" value="$(arg left_sensor_name)"/></include>` could also just be `<group><let name="name" value="$(arg left_sensor_name)"/><include ...></include></group>`
  - including a launch file but not providing the parent scope is a rarer use case (in my opinion), and I personally don't find it surprising that this is the default behavior
- The `Group` action's purpose is to provide scoping and forwarding control for a group of actions, duplicating this behavior in the `IncludeLaunchDescription` action causes unnecessary overlap between the classes
  - what happens when someone updates the behavior for the group action and forgets to do so for the include action? or vice versa? keeping scoping and forwarding in the group action is good for code cohesion (in my opinion)
  - why not drop group and just have scoping and forwarding on include? it is useful to be able to control scoping and forwarding on a group of actions without putting them in a separate launch file
- We can (and should) improve the ergonomics of these workflows and reduce surprise where possible
  - I would prefer to avoid arguments that change the behavior of `IncludeLaunchDescription`, especially ones that overlap with `Group`
  - instead limit changes to the frontend, e.g. `<include scoped=true ...>...</include>` in XML is ok with me for syntactic sugar, but under the hood it would make more sense as a `Group` action that contains an `IncludeLaunchDescription` action
  - if we must have a new Python class, name it differently (use the type system to convey behavior changes, not arguments), e.g. `ScopedIncludeLaunchDescription`, but that class should explicitly be a wrapper for a `Group` action that contains a `IncludeLaunchDescription` instance and provides the frontend syntactic sugar

I think I'll stop there, since this post is already long, but I'm happy to continue the conversation on any of these particular points.

Also, I usually don't have time to join the ROSGraph working group, but if you'd like me to join a specific instance of it to discuss this or even have a dedicated meeting to discuss these issues, let me know. I'm happy to try and arrange that.

**Comment by peci1** [comment]
> * instead limit changes to the frontend, e.g. `<include scoped=true ...>...</include>` in XML is ok with me for syntactic sugar, but under the hood it would make more sense as a `Group` action that contains an `IncludeLaunchDescription` action

I fully agree with this kind of solution.
